### PR TITLE
Channel retention policy overrides for big teams

### PR DESCRIPTION
**Pull request by buoyad**
- [x] match designs (punting on save state)
- [x] hide for non-admins
- [x] add provider to info panel stories

### CODE REVIEW

**Comment by buoyad** [comment]
@keybase/react-hackers this is ready!

File: shared/chat/conversation/info-panel/index.js [hunk 1]
```diff
@@ -89,6 +94,31 @@ const getDividerStyle = (row: DividerRow) => ({
...
+  containerStyle: platformStyles({
+    common: {
+      marginLeft: 16,
+      marginRight: 45,
+    },
+    isMobile: {
+      marginRight: 16,
+    },
+  }),
+  dropdownStyle: platformStyles({
```
Comment (chrisnojima) [inline shared/chat/conversation/info-panel/index.js]: platformStyles not really required here

File: shared/common-adapters/popup-menu.desktop.js [hunk 2]
```diff
@@ -53,8 +53,8 @@ class ModalLessPopupMenu extends Component<ModalLessPopupMenuProps> {
                   className={hoverClassName}
                   style={{...stylesRow, ...styleClickable}}
                   onClick={event => {
-                    i.onClick && i.onClick()
                     this.props.closeOnClick && this.props.onHidden && this.props.onHidden()
+                    i.onClick && i.onClick()
```
Comment (chrisnojima) [inline shared/common-adapters/popup-menu.desktop.js]: why is this necessary

File: shared/constants/chat2/meta.js [hunk 3]
```diff
@@ -208,6 +209,14 @@ export const inboxUIItemToConversationMeta = (i: RPCChatTypes.InboxUIItem) => {
     notificationsMobile,
   } = parseNotificationSettings(i.notifications)
```
Comment (chrisnojima) [inline shared/constants/chat2/meta.js]: maybe better to do 
```
let retentionPolicy
if (isTeam) {
  retentionPolicy = i.convRetention 
    ? serviceRetentionPolicyToRetentionPolicy(i.convRetention) 
    : makeRetentionPolicy({type: 'inherit'})
} else {
  retentionPolicy = makeRetentionPolicy()
}
```

File: shared/constants/chat2/meta.js [hunk 4]
```diff
@@ -308,3 +319,14 @@ export const findConversationFromParticipants = (state: TypedState, participants
       meta.teamType === 'adhoc' && meta.participants.toSet().equals(toFind)
   )
 }
+
+export const getConversationRetentionPolicy = (
+  state: TypedState,
+  conversationIDKey: Types.ConversationIDKey
+) => {
+  const conv = state.chat2.metaMap.get(conversationIDKey)
```
Comment (chrisnojima) [inline shared/constants/chat2/meta.js]: just use Constants.getMeta(conversationIDKey) which returns the blank meta if nothing is there

File: shared/reducers/chat2.js [hunk 5]
```diff
@@ -126,6 +127,19 @@ const metaMapReducer = (metaMap, action) => {
       return ['inboxSyncedClear', 'leftAConversation'].includes(action.payload.reason)
         ? metaMap.clear()
         : metaMap
+    case Chat2Gen.updateConvRetentionPolicy:
+      const update = action.payload.update
+      if (!update.conv) {
```
Comment (chrisnojima) [inline shared/reducers/chat2.js]: why allow this?

File: shared/reducers/chat2.js [hunk 6]
```diff
@@ -126,6 +127,19 @@ const metaMapReducer = (metaMap, action) => {
...
+        logger.warn('Got conv retentionpolicy update with no inboxUIItem')
+        return metaMap
+      }
+      const newMeta = Constants.inboxUIItemToConversationMeta(update.conv)
+      if (!newMeta) {
+        // public conversation, do nothing
```
Comment (chrisnojima) [inline shared/reducers/chat2.js]: this is not the only reason why newMeta can be null, also i don't think you need to enumerate that here

File: shared/teams/team/settings/retention/index.js [hunk 7]
```diff
@@ -206,12 +248,34 @@ const policyToComparable = (p: _RetentionPolicy, parent: ?_RetentionPolicy): num
   }
   return res
 }
-const policyEquals = (p1?: _RetentionPolicy, p2?: _RetentionPolicy): boolean => {
+// For getting the number of days a retention policy resolves to
+const policyToDays = (p: RetentionPolicy, parent?: RetentionPolicy) => {
+  let ret = 0
```
Comment (chrisnojima) [inline shared/teams/team/settings/retention/index.js]: might as well call this 'days'

File: shared/teams/team/settings/retention/index.js [hunk 8]
```diff
@@ -206,12 +248,34 @@ const policyToComparable = (p: _RetentionPolicy, parent: ?_RetentionPolicy): num
...
     return p1.type === p2.type && p1.days === p2.days
   }
   return p1 === p2
 }
 
-export {daysToLabel}
-export default RetentionPicker
+// Switcher to avoid having RetentionPicker try to process nonexistent data
+const RetentionSwitcher = (props: Props) => {
+  if (props.loading) {
```
Comment (chrisnojima) [inline shared/teams/team/settings/retention/index.js]: could just do
```
=> props.loading ? <ProgressIndicator.. /> : <RetentionPicker ... />
```

File: shared/teams/team/settings/retention/warning/container.js [hunk 9]
```diff
@@ -8,17 +8,22 @@ const mapStateToProps = (state: TypedState, {routeProps}) => {
...
-    dispatch(navigateUp())
-  },
-})
+const mapDispatchToProps = (dispatch: Dispatch, {routeProps, navigateUp}) => {
+  const onBack = () => dispatch(navigateUp())
+  return {
+    onBack,
+    onCancel: () => {
+      onBack()
+      const onCancel: () => void = routeProps.get('onCancel')
```
Comment (chrisnojima) [inline shared/teams/team/settings/retention/warning/container.js]: typed as being non null, likely need ? if you're afraid onCancel can be undefined

File: shared/teams/team/settings/retention/warning/container.js [hunk 10]
```diff
@@ -8,17 +8,22 @@ const mapStateToProps = (state: TypedState, {routeProps}) => {
...
+      onCancel && onCancel()
+    },
+    onConfirm: () => {
+      onBack()
+      const cb: () => void = routeProps.get('onConfirm')
```
Comment (chrisnojima) [inline shared/teams/team/settings/retention/warning/container.js]: same here

File: shared/common-adapters/popup-menu.desktop.js (hunk 2 above)
Comment (chrisnojima) [reply shared/common-adapters/popup-menu.desktop.js]: mmmm, wonder what'll break if we make this change. If you use the navigateUp that comes from routeTree it should do nothing if the route has changed

Comment (buoyad) [reply shared/common-adapters/popup-menu.desktop.js]: the issue wasn't a race, but that the callback was doing a `navigateAppend` on the dropdown route when it was only a child of the dropdown's parent. I could change it so the callback navigates up before appending to avoid making this change, but when we figure out a better popup solution I think the contract should be that the dropdown closes before callbacks are called (cc @akalin-keybase, if you're still interested in that ticket).

File: shared/reducers/chat2.js (hunk 5 above)
Comment (buoyad) [reply shared/reducers/chat2.js]: I think this would only happen for old subteams that have no conversations set up; in which case nobody using the client would be able to see any retention controls / try to interact with policies (implicit admins can't). Implicit admins might be getting this notification, though, so I figured it'd be best to silently warn

Comment (buoyad) [reply shared/reducers/chat2.js]: Actually reading over that it might be that implicit admins don't get the notice. @mlsteele do they?

Comment (buoyad) [reply shared/reducers/chat2.js]: nevermind I was thinking about this all wrong. the notification handler should check for `update.conv` and request it if it's not present, and I'll change `updateConvRetentionPolicy` to make the inboxUIItem non-optional

**Comment by buoyad** [comment]
ok @chrisnojima, addressed comments, ptal!
